### PR TITLE
Abstracting data pipline, CNN and RNN pre-processing and fixing syntaxis problems.

### PR DESCRIPTION
**Pull request by imgmaterial**
Majority of the work that was done is getting the code to run again and abstracting various parts of the data processing to allow writing separate scripts for data processing. I removed majority of hard coded paths from the code. A large part of the changes seen in the code is file renames, this was done to allow the files to be easily imported into new scripts.

### CODE REVIEW

File: DataExtractor.py [hunk 1]
```diff
@@ -106,15 +77,15 @@ def Interloper(self):#Interloper is responsible for extracting the Energy and th
...
+                          EDF.to_csv("Directory/Energy{}.csv".format(self.OutputNumber),mode='w',chunksize=5000,index=False)
                           #XDF.to_csv("/Directory/X{}.csv".format(self.OutputNumber),mode='w',chunksize=5000,index=False)
                           #YDF.to_csv("/Directory/Y{}.csv".format(self.OutputNumber),mode='w',chunksize=5000,index=False)
                           #ZDF.to_csv("/Directory/Z{}.csv".format(self.OutputNumber),mode='w',chunksize=5000,index=False)
                           print("CSV: Done")
                  else:
-                          IDF.to_csv("/home/jacoblindahl/EcalID{}.csv".format(self.OutputNumber),mode='a',chunksize=5000,index=False, header=False)#This simply appends the data to the already created file (mode='a')
-                          EDF.to_csv("/home/jacoblindahl/Energy{}.csv".format(self.OutputNumber),mode='a',chunksize=5000,index=False, header=False)
+                          IDF.to_csv("Directory/EcalID{}.csv".format(self.OutputNumber),mode='a',chunksize=5000,index=False, header=False)#This simply appends the data to the already created file (mode='a')
```
Comment (bryngemark) [inline DataExtractor.py]: How do you handle this when you're running the code, are you replacing `Directory` by something else? Consider making the path a variable too, that can be read from an input argument to the script.

File: ETSXRNN_ANN.py [hunk 2]
```diff
@@ -104,13 +104,13 @@ def __getitem__(self, idx) :#This function actually loads the data
 
     
     for file_name in batch_x:#Loads each file with the filenames included 
-        A = np.load('DataDirectory/' + str(file_name))
+        A = np.load('Output/Pre-Processing/RNN/ETSX/' + str(file_name))
```
Comment (bryngemark) [inline ETSXRNN_ANN.py:107]: We can think about paths two ways. Either, this is something the user always chooses for themself. Or, it's a larger structure within the code. In the latter case, there should be some setup steps or directory creation/existence checks inside the code being run. In the former cae, we should let all these be variables that can be set by reading an input variable passed to the script, or parsed from some general configuration file.

File: ETSXRNN_ANN.py [hunk 3]
```diff
@@ -153,16 +153,16 @@ def __getitem__(self, idx) :#This function actually loads the data
 lstm_1 = keras.layers.GRU(128, activation='tanh',return_sequences=True)(masking)#This is the first GRU layer and since return sequences is True this means that it can back propagate with the other GRU layers 
 
 
-lstm_2 = keras.layers.GRU(32, activation='tanh',return_sequences=True,recurrent_dropout=0.3)(lstm_1)#This layer also has recurrent dropout added in the layer 
+lstm_2 = keras.layers.GRU(32, activation='tanh',return_sequences=True,recurrent_dropout=0)(lstm_1)#This layer also has recurrent dropout added in the layer
```
Comment (bryngemark) [inline ETSXRNN_ANN.py:156]: Not sure the comment about the dropout is true anymore?

File: ExtractDataScript.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import pandas as pd
+from DataExtractor import DataExtractor
+csv_path = ["Directory/LIST1.csv","Directory/LIST2.csv","Directory/LIST3.csv","Directory/LIST4.csv"]
+data_path = ["/projects/hep/fs9/shared/ldmx/ldcs/gridftp/mc23/v14/4.0GeV/v3.2.2-1e-v14/",
+              "/projects/hep/fs9/shared/ldmx/ldcs/gridftp/mc23/v14/4.0GeV/v3.2.2-2e-v14/",
+              "/projects/hep/fs9/shared/ldmx/ldcs/gridftp/mc23/v14/4.0GeV/v3.2.2-3e-v14/",
+              "/projects/hep/fs9/shared/ldmx/ldcs/gridftp/mc23/v14/4.0GeV/v3.2.2-4e-v14/"]
+event_count = [1,2,3,4]
+max_length = [155,250,360,450]
```
Comment (bryngemark) [inline ExtractDataScript.py:7]: As we discussed, consider making them all fixed length

File: AuroraJobEx.sh [hunk 5]
```diff
@@ -5,4 +5,4 @@
 #SBATCH --mem-per-cpu=6000 #This tells the system how much memory should be used per CPU involved in the calculation of the job 
 #SBATCH -N 2 #This specifies how many nodes should be used for the job
```
Comment (bryngemark) [inline AuroraJobEx.sh:6]: Do you need two full nodes (40 cores) to be reserved for your job? I would have thought you could do single-core, or is this multithreaded in some way?

**Review by bryngemark — COMMENTED** [review]
General comment: there is one python file per multiplicity. Looking superficially at it, they seem to run the same thing, with some array lengths differing, and likely some indices too? How difficult would it be to make this one single file, with those differences handled by input arguments to a function?

The same comment applies to the repeated loops in many of the files, these might be extracted into separate functions (could just be defs in the same file) and the multiplicity could be an input argument. I think this would make the code more readable and less error prone (avoids copy/paste errors)

**Comment by imgmaterial** [comment]
Thank you for the review! I have addressed some but far from all of the problems in the last contributions. I have added the output path as a parameter to the DataExtractor class. The pre-processing for Ecal part of both the CNN and RNN has been condensed into two files. Along with that I tried to make it more readable and fixed a bug that resulted in not all events being saved after pre-processing. 
After diving a bit more into the code, the problem with differing length has already been addressed in the training scripts as RNN training pads the events with extra zeros before sending them to training.  
Most of the repeated loops are a part of processing scripts working with trigger scintillator information so I will condense those when the changes to ldmx-sw become available.

**Review by bryngemark — COMMENTED** [review]
This looks a lot smarter now, with the condensed functions, and easier to work with.

File: ExtractDataScript.py (hunk 4 above)
Comment (bryngemark) [reply ExtractDataScript.py:7]: dealt with later in the training code

**Comment by bryngemark** [comment]
are you ready to have this merged in for now or are there more things you are still working on?

**Comment by imgmaterial** [comment]
It can be merged in for now

File: AuroraJobEx.sh (hunk 5 above)
Comment (bryngemark) [reply AuroraJobEx.sh:6]: I'll merge this in now, but this question still stands; just want to check that you aren't requesting more resources than you nee since that leads to unnecessary waiting time for both you and others
